### PR TITLE
Kifi installation

### PR DESCRIPTION
**Pull request by andrewconner**

### CODE REVIEW

**Comment by andrewconner** [comment]
@2is10 Please review & merge

File: shoebox/conf/routes [hunk 1]
```diff
@@ -29,8 +29,7 @@ GET     /users/slider/updates       com.keepit.controllers.CommentController.get
 GET     /users/keepurl              com.keepit.controllers.UserController.usersKeptUrl(url: String)
 GET     /users/friends              com.keepit.controllers.UserController.getSocialConnections()
 
-GET     /search2                    com.keepit.controllers.SearchController.search2(term: String, maxHits:Int, lastUUID: Option[String], context: Option[String])
-GET     /search                     com.keepit.controllers.SearchController.search(term: String, maxHits:Int, lastUUID: Option[String], context: Option[String])
+GET     /search                     com.keepit.controllers.SearchController.search(term: String, maxHits:Int, lastUUID: Option[String], context: Option[String], kifiVersion: String ?= "")
```
Comment (2is10) [inline shoebox/conf/routes]: Can we make it an Option[KifiVersion]? Best to let the framework do instantiation from request parameters. Makes our code smaller and clearer.

File: shoebox/conf/evolutions/shoebox/20.sql [hunk 2]
```diff
@@ -1,6 +1,6 @@
 # --- !Ups
 
-CREATE TABLE extension_version (
+CREATE TABLE kifi_installation (
```
Comment (2is10) [inline shoebox/conf/evolutions/shoebox/20.sql:3]: please also rename browser_instance_id => external_id

File: shoebox/app/com/keepit/model/KifiInstallation.scala [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+  }
+
+}
+
+private[model] class KifiInstallationEntity extends Entity[KifiInstallation, KifiInstallationEntity] {
+  val createdAt = "created_at".JODA_TIMESTAMP.NOT_NULL(currentDateTime)
+  val updatedAt = "updated_at".JODA_TIMESTAMP.NOT_NULL(currentDateTime)
+  val userId = "user_id".ID[User].NOT_NULL
+  val version = "version".VARCHAR(16).NOT_NULL
+  val externalId = "browser_instance_id".EXTERNAL_ID[KifiInstallation].NOT_NULL
```
Comment (2is10) [inline shoebox/app/com/keepit/model/KifiInstallation.scala]: browser_instance_id => external_id

File: shoebox/app/com/keepit/model/KifiInstallation.scala [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+      new UserAgent(userAgent)
+    }
+}
+
+case class KifiVersion(major: Int, minor: Int, patch: Int = 0, tag: String = "") extends Ordered[KifiVersion]  {
+  assert(major >= 0 && minor >= 0 && patch >= 0)
+  def compare(that: KifiVersion) =
+    (this.major - that.major) * 10000 +
+    (this.minor - that.minor) * 100 +
+    (this.patch - that.patch)
```
Comment (2is10) [inline shoebox/app/com/keepit/model/KifiInstallation.scala:41]: more efficient:

(this.major - that.major) << 20 +
(this.minor - that.minor) << 10 +
(this.patch - that.patch)

File: shoebox/app/com/keepit/model/KifiInstallation.scala [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+      val v = t(0).split('.')
+      assert(v.length > 1)
+      val tag = if(t.length > 1) t(1) else ""
+      KifiVersion(safeVersion(v,0), safeVersion(v,1), safeVersion(v,2), tag)
+    } catch {
+      case _ =>
+        log.warn("Invalid kifi version: %s".format(version))
+        KifiVersion()
+    }
+  }
```
Comment (2is10) [inline shoebox/app/com/keepit/model/KifiInstallation.scala:65]: a regexp could shorten this function to 1 or 2 lines

`"""(\d+)\.(\d+)(?:\.(\d+))?(?:-([a-zA-Z0-9-])+)?""".r`

File: shoebox/app/com/keepit/model/KifiInstallation.scala [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+      val t = version.split('-')
+      assert(t.length > 0)
+      val v = t(0).split('.')
+      assert(v.length > 1)
+      val tag = if(t.length > 1) t(1) else ""
+      KifiVersion(safeVersion(v,0), safeVersion(v,1), safeVersion(v,2), tag)
+    } catch {
+      case _ =>
+        log.warn("Invalid kifi version: %s".format(version))
+        KifiVersion()
```
Comment (2is10) [inline shoebox/app/com/keepit/model/KifiInstallation.scala:63]: better to throw

File: chrome/manifest.json [hunk 7]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "Kifi Beta",
-  "version": "2.0",
+  "version": "2.0.1-alpha",
```
Comment (2is10) [inline chrome/manifest.json]: do we need to remember to change this temporarily each time we do a new release?

Comment (andrewconner) [inline chrome/manifest.json]: We could just increment to the next one, and don't worry about tags?

File: shoebox/app/com/keepit/model/KifiInstallation.scala (hunk 6 above)
Comment (eishay) [inline shoebox/app/com/keepit/model/KifiInstallation.scala:63]: agree, never swallow coding errors

**Pull request merged**